Implement pcntl_waitid

vrza opened the pull request:
Closes  #14616

To round out the support for the "wait" family of functions this PR adds the `pcntl_waitid` function.

Here's the POSIX reference to `waitid`:

https://pubs.opengroup.org/onlinepubs/9699919799/functions/waitid.html

> The waitid() function shall obtain status information pertaining to termination, stop, and/or continue events in one of the caller's child processes

Since status information struct is shared with `sigwaitinfo`, relevant code from `pcntl_sigwaitinfo` has been reused.

Note to maintainers: if this patch is accepted, I'll be happy to provide the [English documentation in the follow-up PR](https://github.com/php/doc-en/pull/3518).

devnexen commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +385,57 @@ PHP_FUNCTION(pcntl_waitpid)
+
+	// check idtype is one of P_PID, P_PIDFD, P_PGID, P_ALL
+	if (idtype != P_ALL && idtype != P_PID && idtype != P_PIDFD && idtype != P_PGID) {
+		php_error_docref(NULL, E_WARNING, "idtype must be one of P_PID, P_PIDFD, P_PGID, P_ALL");
```
nit: preferably throws a ValueError exception instead.

devnexen commented on ext/pcntl/pcntl.stub.php:
```diff
@@ -25,6 +25,49 @@
+ * @var int
+ * @cvalue LONG_CONST(P_PIDFD)
+ */
+const P_PIDFD = UNKNOWN;
```
that s a linux extension.

vrza replied:
Good catch, thanks. I added a #ifdef __linux__ as guard there.

devnexen replied:
nope. ifdef P_PIDFD is better. see the man page. old kernels do not support it and we deal with centos 7/8 users.

vrza replied:
Good to know. I did originally try checking for the constant itself, but for some reason it doesn't work for me, preprocessor seems to not see the constant and skips the whole ifdef block block:

```
 #ifdef P_PIDFD
/**
 * @var int
 * @cvalue LONG_CONST(P_PIDFD)
 */
const P_PIDFD = UNKNOWN;
#endif
```

devnexen replied:
I see, maybe it s because of [this inclusion](https://github.com/php/php-src/blob/e23bc2d50c3663d0c04a02902223ba48b9932df5/ext/pcntl/pcntl.c#L38). But regardless I ll test your branch locally later tonight. cheers.

vrza replied:
Quick smoke test:

```
php_error_docref(NULL, E_WARNING, "P_PIDFD is defined as: %d", P_PIDFD);                                                
#ifndef P_PIDFD                                                                                                         
php_error_docref(NULL, E_WARNING, "P_PIDFD is not defined");                                                            
#endif  
```

yields

```
Warning: pcntl_waitid(): P_PIDFD is defined as: 3 in /home/random/code/php/waitid/waitid.php on line 32

Warning: pcntl_waitid(): P_PIDFD is not defined in /home/random/code/php/waitid/waitid.php on line 32
```

vrza replied:
Looks like these constants are defined in `<linux/wait.h>` so would have to pull that header in first (but only if on Linux) in order to check them.

vrza commented:
@devnexen Thanks for the quick review! I addressed your comments (ValueErrors, unit test, guard for the Linux speciifc P_PIDFD), please take another look.

devnexen commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +385,55 @@ PHP_FUNCTION(pcntl_waitpid)
+		Z_PARAM_LONG(options)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (idtype != P_ALL && idtype != P_PID && idtype != P_PGID && idtype != P_PIDFD) {
```
last thing: do not forget what I mentioned about non standard flags. You may develop under a recent Linux distro but not the case for everyone else. macOs users, BSD ... you need to see the multiplatform context. Or if you do not want, you can just deal with only standard flags and someone else will deal with other oses, no big deal :).

vrza replied:
Agreed, both idtype and flags are platform-specific.

It seems that the best course is to leave validation to the `waitid` call itself and let it set `errno`.

devnexen commented:
I see the progress but take your time :) I ll have a look later.

devnexen commented on ext/pcntl/pcntl.c:
```diff
@@ -125,6 +125,16 @@ static zend_class_entry *QosClass_ce;
 
 #define LONG_CONST(c) (zend_long) c
 
+#if defined(__linux__)
```
good intention but not quite. see musl define all [here](https://git.musl-libc.org/cgit/musl/tree/include/sys/wait.h#n17).

what you can do however is to add the header presence check in ext/pcntl/config.m4

AC_CHECK_HEADER([linux/wait.h])
which should create the following preprocessor

HAVE_LINUX_WAIT_H

(so before recompiling you need to start fresh including doing ./buildconf)

After that, you should be able to test P_PIDFD.

vrza replied:
Good point and good idea, made the change.

devnexen commented on ext/pcntl/pcntl.c:
```diff
@@ -125,6 +125,16 @@ static zend_class_entry *QosClass_ce;
+#include <linux/wait.h>
+#endif
+#if defined (P_ALL)
+#define HAVE_P_ALL 1
```
supporting waitid includes having P_ALL, P_PID, P_PGID guaranteed.

vrza replied:
I did some investigation and broke them up into:

- POSIX (incl. OpenBSD and Solaris)
- Linux (POSIX + P_PIDFD)
- NetBSD (POSIX + P_UID + P_GID + P_SID)
- FreeBSD (NetBSD + P_JAILID

vrza replied:
On the other hand macOS seems to be using an antiquated version of FreeBSD libc, perhaps it is a POSIX-like system that doesn't have waitid and thus this function should be optional in pcntl?

devnexen replied:
Yes good point I forgot to mention it but that s easy enough, you need to add waitid to the list of existing check in ext/pcntl/config.m4 then guarding things (including in the stub file) with HAVE_WAITID.

vrza replied:
Yeah, that was pretty straightforward. Done.

devnexen commented on ext/pcntl/tests/pcntl_waitid.phpt:
```diff
@@ -0,0 +1,30 @@
+if ($pid == -1) {
+    die("failed");
+} else if ($pid) {
+    $status = 0;
```
once you re done we need to test pcntl_waitid also with incongruous values on purpose.

devnexen commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +401,42 @@ PHP_FUNCTION(pcntl_waitpid)
 }
 /* }}} */
 
+/* {{{ Waits on or returns the status of a forked child as defined by the waitid() system call */
```
nit: no need to bother with these with new functions, you could not know but we no longer need these.

devnexen commented on ext/pcntl/tests/pcntl_waitid.phpt:
```diff
@@ -0,0 +1,41 @@
+    die("failed");
+} else if ($pid) {
+    // test invalid arguments
+    var_dump(@pcntl_waitid(-2, $pid, $siginfo, WSTOPPED));
```
would be nice to see related error messages also trying bad values for last argument.

Girgias replied:
I agree, I would not suppress the warnings here. Worst case we can standardize them ourselves depending on what the errno is https://pubs.opengroup.org/onlinepubs/7908799/xsh/waitid.html

vrza replied:
Added checks for error messages, thanks.

@devnexen Bad value for last argument test was already there (WNOHANG by itself is a bad value for flags, a one of WEXITED, WSTOPPED, WCONTINUED is mandatory). I added a comment to the test as this might not be obvious.

devnexen replied:
Sure but I was thinking about something more "silly" e.g. `PHP_INT_MAX`.

vrza replied:
Okay, added a couple of those as well.

Girgias commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +403,41 @@ PHP_FUNCTION(pcntl_waitpid)
+	zend_long id = 0;
+	zval *user_siginfo = NULL;
+	zend_long options = WEXITED;
+	int success;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_LONG(idtype)
+		Z_PARAM_LONG(id)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(user_siginfo)
+		Z_PARAM_LONG(options)
+	ZEND_PARSE_PARAMETERS_END();
+
+	errno = 0;
+	siginfo_t siginfo;
+
+	success = waitid((idtype_t) idtype, (id_t) id, &siginfo, (int) options);
```
I would merge the declaration and assignment and rename the variable, as `success` is not very accurate.
```suggestion

	ZEND_PARSE_PARAMETERS_START(2, 4)
		Z_PARAM_LONG(idtype)
		Z_PARAM_LONG(id)
		Z_PARAM_OPTIONAL
		Z_PARAM_ZVAL(user_siginfo)
		Z_PARAM_LONG(options)
	ZEND_PARSE_PARAMETERS_END();

	errno = 0;
	siginfo_t siginfo;

	int status = waitid((idtype_t) idtype, (id_t) id, &siginfo, (int) options);
```

Girgias commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +403,41 @@ PHP_FUNCTION(pcntl_waitpid)
+	int success;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_LONG(idtype)
```
There should be a check + ValueError for the valid type.

vrza replied:
Strict idtype validation is platform-specific.

Here are the idtype definitions for platforms I checked:

https://github.com/torvalds/linux/blob/master/include/uapi/linux/wait.h#L16
https://github.com/openbsd/src/blob/master/sys/sys/wait.h#L87
https://github.com/NetBSD/src/blob/trunk/sys/sys/idtype.h#L42
https://cgit.freebsd.org/src/tree/sys/sys/wait.h#n93
https://github.com/illumos/illumos-gate/blob/b41e2fb6b2658a36accc73b3b86d99375e700dba/usr/src/uts/common/sys/procset.h#L58

Let's take NetBSD as a particular platform where valid idtypes are -1 <= idtype <= 15, with an additional value _P_MAXIDTYPE = 0x7fffffff (the  highest 32-bit two's complement representable positive integer). -1 on NetBSD translates to current process effective PID/GID.

I see four possible design choices here:

1. Leave idtype validation to the underlying C library function. This is what I would do, because what constitutes valid idtypes is platform-specific.
2. Implement an (ifdef) case-switch of various platform-specific validations. This could be a burden for long term maintenance, as ports to new platforms are introduced, and even existing implementations can change.
3. Implement some sort of weak validation, e.g. idtype >= -1. The downside is that such validation not catch all invalid values, and such an assumption could become incorrect on some platform in the future.
4. Support only POSIX-defined idtypes, regardless of the platform. This is clean to implement but users might find it overly restrictive.
                                             
Thoughts and suggestions are welcome.

Girgias replied:
We generally do option 2 in other extensions.

But having 3 is also fine, it does not need to necessarily catch *all* issues, but erroring early with a ValueError for obviously wrong values is good.

vrza replied:
Ok, added weak validation per option 3 for now.

Another option would be better to check for EINVAL in the function itself, and convert it to ValueError. Not sure if this is consistent with the surrounding parts of the API.

vrza replied:
There is also a question of API consistency, other functions in the family (`pcntl_wait` and `pcntl_waitid`) set errno and don't throw ValueErrors, so having `pcntl_waitid` also throw ValueErrors would not be consistent with the rest of the package.

@devnexen would love to hear your thoughts on API design consistency, perhaps there is context I'm missing, and also of the idea of redundant checks vs converting EINVAL errnos to ValueError throwables.

devnexen replied:
in fact, we should strive for better so other existing functions should throw ValueError too in further improvements.

vrza replied:
Sounds good, agreed.

Changed code to throw a ValueError when invalid arguments are passed.

Girgias commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +403,41 @@ PHP_FUNCTION(pcntl_waitpid)
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_LONG(idtype)
+		Z_PARAM_LONG(id)
```
Should check if this is a valid ID (I would guess an unsigned int) for the modes that require them.

I would probably also make this argument nullable for the `P_ALL` mode, as it seems it doesn't care about the ID: https://pubs.opengroup.org/onlinepubs/7908799/xsh/waitid.html

vrza replied:
Good point about making the argument nullable, thanks.

Regarding validation check, the discussion is similar to idtype validation -- what constitutes a valid id depends on platform and idtype. Let's first decide how to proceed with idtype validation, and then we can come back to decide on id validation.

vrza commented:
It seems that autotools detect `waitid`  as present on macOS -- although undocumented it's declared in their headers.

Not sure if that means it works on macOS as expected, at least the unit test seems to hang.

Girgias commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +403,45 @@ PHP_FUNCTION(pcntl_waitpid)
+	errno = 0;
+	siginfo_t siginfo;
+
+	int status = waitid((idtype_t) idtype, (id_t) id, &siginfo, (int) options);
+
+	if (status == -1) {
+		if (errno == EINVAL) {
+			zend_value_error("An invalid value was specified for options, or idtype and id specify an invalid set of processes");
+			RETURN_THROWS();
+		}
+		PCNTL_G(last_error) = errno;
+		php_error_docref(NULL, E_WARNING, "%s", strerror(errno));
+		RETURN_FALSE;
+	}
```
The value errors should be triggered *prior* to the call to waitid IMHO.

Moreover, if you pass `null` for the ID and the option expects an actual int ID this should also throw.

vrza replied:
@Girgias If you pass `null` for the id and the `idtype` requires an actual ID, the function will throw.

Please elaborate why you think value errors should be thrown before the call to `waitid`.

devnexen replied:
it is for value check quality control before doing the call so we do not pass know range of invalid values.

vrza replied:
If we pass invalid values, [POSIX standard mandates](url) that the system API implementation should fail with EINVAL. Thus, the underlying implementation is mandated to perform the validation for us.

Duplicating the validation code by maintaining a copy of it in the PHP standard library would not be in line with the [Unix software philosophy](https://en.wikipedia.org/wiki/Unix_philosophy#Eric_Raymond's_17_Unix_Rules). Such a design would not follow principles of modularity (or [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)), composition, Bertrand Meyer's [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), [single source of truth](https://en.wikipedia.org/wiki/Single_source_of_truth) etc.

devnexen replied:
You can have a look at the posix extension, e.g. posix_mknod, there is some pre-validations.

vrza replied:
I never copy any code or software design without thoroughly reviewing it and understanding what it does and why.

@devnexen  You sent me on an errand to read and review the implementation of `posix_mknod`, and it took me at least an hour (reading the POSIX docs, reading the man pages, checking underlying implementations on different platforms) to do so. As I will elaborate shortly, I found this particular errand to be a waste of time. The first question that comes to mind is: did you thoroughly review the function you referenced (`posix_mknod`) yourself at some point, and found that it is relevant and of exemplary design?

Now to elaborate, by reviewing the code of `posix_mknod`. It does three checks:

1. Checks if the `open_basedir` `php.ini` setting restricts access to the path where user wants to create the file. This check is legitimate, but completely irrelevant to `pcntl_waitid`, as `waitid` has nothing to do with the file system (hence `open_basedir` restrictions are not applicable).

2. If checks if the user is attempting to create a block or character device file with a major number of 0. First thing to notice here is that major and minor device numbers are [not portable per POSIX](https://pubs.opengroup.org/onlinepubs/009696699/functions/mknod.html). But most importantly, I see no reason why PHP should prevent the user from create a device file with major number 0?! For example, in OpenBSD, /dev/console is major 0, minor 0, so at least on one system PHP will prevent the user from performing a perfectly legitimate action! Am I missing something, do you feel that this check is legitimate and relevant?

3. It checks for the presence of the `makedev(3)` utility macro. `makedev` macro implementation(s) could've been duplicated in-line, but the `posix_mknod` developer chose not to do so, instead requiring requiring the macro to be present if the user is creating a S_IFCHR or S_IFBLK file. This example reinforces my argument that you should not duplicate existing system functionality in PHP standard library, and does so very strongly because `makedev` is not standardizes or mandated by POSIX! Yet PHP will refuse to create the dev param itself if libc macro is not present!  `pcntl_waitid` itself of course does not need any aditional functions or macros, so this check is relevant only as an example showing that it is perfectly fine not to duplicate system code.

4. Finally, if mknod fails, it exposes the errno to the user (without converting it to a ValueError, tho). This is legitimate, and this is similar to waitid does.

To summarize, in addition to checking errno, posix_mknod makes one checks that is  completely irrelevant to our discussion (the php.ini restriction feature), one check that avoids adding code to PHP (relying on a utility macro), one check (major == 0) that I would consider problematic: e.g. on OpenBSD it's an outright bug. In short, I see nothing relevant to learn from posix_mknod code that we could use to improve pcntl_waitid code.

devnexen replied:
It was an example, for sure not the same case .. there is [gd](https://github.com/php/php-src/blob/678a481f62db7776156bc69eaa87aaf2ec9e6673/ext/gd/gd.c#L1956) here when we control the quality before the call.

devnexen replied:
Also part of it is in some platforms, a call does not necessarily fully follow the protocol or some implementations are buggy.

vrza replied:
> there is [gd](https://github.com/php/php-src/blob/678a481f62db7776156bc69eaa87aaf2ec9e6673/ext/gd/gd.c#L1956) here when we control the quality before the call.

And there's a good reason for that -- the gd API is poorly designed in the sense that the export function [gdImagePngEx](https://libgd.github.io/manuals/2.2.2/files/gd_png-c.html#gdImagePngEx) will neither do input validation/sanitization, nor report back an error to the user. It will pass it to libpng, which will in turn pass it to zlib, but not before creating the file and outputting the PNG header. zlib will validate the compression level and fail and user will end up with a broken png file, consisting of only the PNP header (no zlib data chunks as the program failed on first call to zlib).

PHP dev decided to throw an error to the user where the underlying implementation would just produce some stderr messages + a broken file. Absolutely agree with the choice there.

In the gd example, PHP is adding input validation because it validation is not part of the direct underlying API (it happens many stack frames below), so it is much easier and more elegant than having to grep through own stderr and delete failed export files.

In our waitid example, validation is mandated by the underlying API, and validation results are readily available through said API.

It's a very different case.

If you can  point out, very concretely, at particular buggy or non-standard implementations of `waitid(2)`, and how they should be addressed, in as much technical detail as I am offering in my explanations, I'll be happy to make such concrete changes. If not, then let's please stop this licitation of random code examples that are not relevant to the code at hand. They only distract us from finding actual existing issues with the PR.

vrza commented:
After some testing on FreeBSD, it seems that, while on Linux P_ALL and friends are preprosessor `#defines`, on other platforms (like FreeBSD) they are be parts of an enum, and thus not visible to the preprocessor.

vrza commented:
Please review this patch, it is now done and tested to work on Linux, FreeBSD, and macOS.

On macOS, though, existence of `waitid(2)` is undocumented, but it seems to work for the POSIX-compliant tested use cases.

Error handling is done by accessing `errno` through `pcntl_get_last_error`, and existing constants `PCNTL_ECHILD`, `PCNTL_EINTR` and `PCNTL_EINVAL`. This is in line what POSIX wrapper functions in PHP generally do, in particular the implementation is consistent with close relatives `pcntl_wait` and `pcntl_waitpid` which I reviewed and found to be of satisfactory design.

devnexen commented:
The FreeBSD failing test seems legit.

devnexen commented:
@petk what do you think build system wise ?

vrza commented:
> The FreeBSD failing test seems legit.

@devnexen  Right, there was a race condition in the test. Fixed it, thanks for noticing and raising.

On a related note, in multiple other unit tests there are guards that skip them on FreeBSD. I suspect many of them could also be due to incorrect (buggy) test code.

For example, there is [an issue in pcntl/tests/002.phpt](https://github.com/php/php-src/blame/54e5e7b507c15e29f2fc64c64047ded68283576a/ext/pcntl/tests/002.phpt#L15), that was silenced by Nikita Popov in 2021:

```
74859783398 (Nikita Popov       2021-06-11 11:57:42 +0200 15) elseif (str_contains(PHP_OS, 'FreeBSD')) die('skip Results in parallel test runner hang on FreeBSD');
```

I took a quick look at that test code, looks like a race condition (SIGTERM can be sent too early). I could fix that one as well, but as a separate issue/PR.

petk commented:
> @petk what do you think build system wise ?

Build system changes look very good and professional. Even nicely checked those `HAVE_DECL_` ones with `#if defined` 🤝 Except maybe those two #include arguments can be wraped into quotes without new lines in them: `[#include <sys/wait.h>]`

vrza commented:
@petk Thanks for reviewing 🤝 The aim is to be strict and explicit about requirements, but flexible about the way they are provisioned.

Regarding m4 formatting... Starting with a disclaimer that this is purely a cosmetic discussion with no impact to functionality.. Some of those m4 lines are very long, I'd like to break them up somewhere for readability, popping the includes out seemed like a nice place to break the line, although the list of declarations is also rather long. Eventually includes turned out to be be an short unprotected one-liner.

Perhaps a compromise, with Lisp-style trailing braces:

```
  AC_CHECK_FUNCS([WIFCONTINUED],,
    [AC_CHECK_DECL([WIFCONTINUED], [AC_DEFINE([HAVE_WIFCONTINUED], [1])],,
      [#include <sys/wait.h> ])])

  AC_CHECK_DECLS([WCONTINUED, WEXITED, WSTOPPED, WNOWAIT, P_ALL, P_PIDFD, P_UID, P_JAILID],,,
    [#include <sys/wait.h>])
  ```

petk commented:
> @petk Thanks for reviewing 🤝 The aim is to be strict and explicit about requirements, but flexible about the way they are provisioned.
> 
> Regarding m4 formatting... Starting with a disclaimer that this is purely a cosmetic discussion with no impact to functionality... Some of those m4 lines are very long, I'd like to break them up somewhere for readability, popping the includes out seemed like a nice place to break the line, although the list of declarations is also rather long. Eventually includes turned out to be a short, unprotected, one-liner.
> 
> Perhaps a compromise, with Lisp-style trailing braces:
> 
> ```
>   AC_CHECK_FUNCS([WIFCONTINUED],,
>     [AC_CHECK_DECL([WIFCONTINUED], [AC_DEFINE([HAVE_WIFCONTINUED], [1])],,
>       [#include <sys/wait.h>])])
> 
>   AC_CHECK_DECLS([WCONTINUED, WEXITED, WSTOPPED, WNOWAIT, P_ALL, P_PIDFD, P_UID, P_JAILID],,,
>     [#include <sys/wait.h>])
> ```

Perfect. Yes that's how I was thinking also. :D

devnexen commented:
Looking ok by me, cc @kocsismate

vrza commented:
While this is in review I started working on [documentation](https://github.com/php/doc-en/pull/3518).

bukka commented:
Looks reasonable

devnexen reviewed (approved):
Since it targets master, we can always improve/fix upon if needs be. I ll merge it in few days if no disagreement arise in the meantime.

kocsismate commented on ext/pcntl/pcntl.c:
```diff
@@ -385,6 +400,39 @@ PHP_FUNCTION(pcntl_waitpid)
+	ZEND_PARSE_PARAMETERS_START(0, 4)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG(idtype)
+		Z_PARAM_LONG_OR_NULL(id, id_is_null)
```
why is this parameter nullable? Apparently, it's not used anywhere.

vrza replied:
It's a minor style point. From the API design standpoint it's not strictly necessary to allow `id = null`, but since the `id` parameter will be ignored in the characteristic case of `idtype = P_ALL`, it can allow PHP programmers to better express their intent and help readability.

For example, compare:

```php
pcntl_waitid(P_ALL, null, $info);
```

with

```php
pcntl_waitid(P_ALL, 0, $info);
```

or even

```
```php
pcntl_waitid(P_ALL, 42, $info);
```

They all do the same thing, but IMHO the first version best communicates to the person reading the PHP code that we don't really care about the value of the second param (`id`).

In the C code, the `id_is_null` value is not used anywhere, as passing 0 to the system call will suffice, but I'm not sure if it's possible to allow the value to be null without using this Z_PARAM_LONG_OR_NULL macro.

kocsismate commented:
Thanks for the elaborate answer! I support the decision fully, it just looked weird at the first sight that `id_is_null` was unused, but after thinking about it a little bit, we have plenty of similar parameters in php-src where `0` behaves like `null`. So I'm sorry about the noise!

vrza commented:
No problem, thanks for taking the time to review the code. There's a follow up PR, that `[adds a non-POSIX rusage parameter](https://github.com/php/php-src/pull/15921)` on systems that support it, that could also benefit from some additional eyeballs on it.
